2.2-taptweak.ipynb

```diff
@@ -210,37 +217,101 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Key pair generation\n",
-    "privkey1, pubkey1 = generate_key_pair()\n",
-    "print(\"Private key: {}\\nPublic key: {}\\n\".format(x.secret, pubkey.get_bytes().hex()))\n",
+    "# Alice modifies the contract and produces an alternative tweak\n",
+    "alternative_contract = \"Alice pays 0.1 BTC to Bob\"\n",
+    "t2 = sha256(alternative_contract.encode('utf-8'))\n",
+    "print(\"Tweak from original contract: {}\".format(t.hex()))\n",
+    "print(\"Tweak from modified contract: {}\\n\".format(t2.hex()))\n",
+    "\n",
+    "# Alice modifies her original secret and public key\n",
+    "# x` = x - t' + t\n",
+    "x_int = x_key.as_int()\n",
+    "t_int = int.from_bytes(t, \"big\") \n",
+    "t2_int = int.from_bytes(t2, \"big\") \n",
+    "x2_int = (x_int - t2_int + t_int) % SECP256K1_ORDER\n",
+    "x2_key = ECKey().set(x2_int, True)\n",
+    "P2_key = x2_key.get_pubkey()\n",
+    "\n",
+    "# The resulting tweaked public Q key remains the same\n",
+    "Q2_key = P2_key.tweak_add(t2)\n",
+    "print(\"Tweaked public key from original tweak t: {}\".format(Q_key.get_bytes().hex()))\n",
+    "print(\"Tweaked public key from modified tweak t2: {}\\n\".format(Q2_key.get_bytes().hex()))\n",
+    "\n",
+    "# So Alice can still produce a valid signature for Q\n",
+    "msg2 = sha256(b'I agree to the committed contract')\n",
+    "sig2 = q_key.sign_schnorr(msg2)\n",
+    "\n",
+    "# Given P2, Q and the modified contract, \n",
+    "# Alice can demonstrate the following to invalidate the original contract:\n",
+    "\n",
+    "# 1) The modified contract 'Alice pays 0.1 BTC to Bob' is appears to be committed to public key Q\n",
+    "print(\"'Alice pays 0.1 BTC to Bob' appears to be committed to Q: {}\"\\\n",
+    "      .format(P2_key.tweak_add(sha256(\"Alice pays 0.1 BTC to Bob\".encode('utf-8'))) == Q_key))\n",
+    "\n",
+    "# 2) Alice is still the owner of the public key Q = P2 + t2*G\n",
+    "print(\"Alice has produced a valid signature for Q: {}\".format(Q_key.verify_schnorr(sig2, msg)))"
```

Comment:
Might want to summarize what happened and the problem, thus leading into the next section

Reply:
Looks good!